src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -388,6 +388,10 @@ infer' (IfThenElse cond th el) = do
 infer' (Let ds val) = do
   (ds', val'@(TypedValue _ _ valTy)) <- inferLetBinding [] ds val infer
   return $ TypedValue True (Let ds' val') valTy
+infer' (Proxy ty) = do
```

Comment:
Not sure if this branch is correct. Also, I think this could be the source of the infinite loop.